src/fhirtypes/StructureDefinition.ts

```diff
@@ -158,26 +157,75 @@ export class StructureDefinition {
+   * on its ID and sliceName.  This should be used rather than pushing directly to the elements array.
    * @param {ElementDefinition} element - the ElementDefinition to add
    */
   addElement(element: ElementDefinition) {
-    let i = 0;
-    let lastMatchId = '';
-    for (; i < this.elements.length; i++) {
-      const currentId = this.elements[i].id;
-      if (new RegExp(`^${escapeRegExp(currentId)}[.:/]`).test(element.id)) {
-        lastMatchId = currentId;
-      } else if (
-        !new RegExp(`^${escapeRegExp(lastMatchId)}[./:]`).test(currentId) ||
-        // If element is not a slice at this level, and the currentId is a slice, break to add children before slices
-        (new RegExp(`^${escapeRegExp(lastMatchId)}[.]`).test(element.id) &&
-          new RegExp(`^${escapeRegExp(lastMatchId)}[:/]`).test(currentId))
-      ) {
-        break;
+    element.structDef = this;
+    this.addElementToTree(element);
+    if (element.treeParent?.treeChildren != null) {
+      if (element.sliceName != null) {
+        // a newly added element with a slicename should have at least one sibling: its sliced element
+        // all slices come after all child elements.
+        // so, insert a slice before the sliced element's next non-slice sibling that comes after it
+        // this is unfortunately most easily done with string tests since we need to keep slices and reslices organized
+        const slicedElement = element.slicedElement();
+        let i = slicedElement ? this.elements.indexOf(slicedElement) : 0;
+        let lastMatchId = this.elements[i].id;
+        for (; i < this.elements.length; i++) {
+          const currentId = this.elements[i].id;
+          if (new RegExp(`^${this.escapePath(currentId)}[.:/]`).test(element.id)) {
+            lastMatchId = currentId;
+          } else {
+            const lastMatchRegex = this.escapePath(lastMatchId);
+            if (
+              !new RegExp(`^${lastMatchRegex}[./:]`).test(currentId) ||
+              // If element is not a slice at this level, and the currentId is a slice, break to add children before slices
+              (new RegExp(`^${lastMatchRegex}[.]`).test(element.id) &&
+                new RegExp(`^${lastMatchRegex}[:/]`).test(currentId))
+            ) {
+              break;
+            }
+          }
+        }
+        this.elements.splice(i, 0, element);
+      } else {
+        // if this is the only child, splice in after parent
+        // if this is not the only child, splice in after your next-older sibling's last child
+        if (element.treeParent.treeChildren.length === 1) {
+          const parentIndex = this.elements.indexOf(element.treeParent);
+          this.elements.splice(parentIndex + 1, 0, element);
+        } else {
+          const olderSibling = element.treeParent.treeChildren.slice(
+            -2,
+            element.treeParent.treeChildren.length - 1
+          )[0];
```

Comment:
Why is -2 the start index here?

Reply:
This slice gets us the second-to-last element of `element.treeParent.treeChildren`. Since `element` has already been added to the tree, it is the last element of `element.treeParent.treeChildren`. So, the second-to-last element is the sibling that needs to be checked for any child elements.

Reply:
Ahhh okay, that makes sense. I think part of why I got confused was that the `slice` used `-2` and `element.treeParent.treeChildren.length - 1`. Is there a reason to no just use `-2` and `-1`? Or `element.treeParent.treeChildren.length - 2` and `element.treeParent.treeChildren.length - 1`? Or am I just messing up my math completely?

Reply:
Oh, it's because I forgot that the second argument can also be relative to the end of the array, so this can just be `slice(-2, -1)`. That's much simpler.